onnxruntime/core/framework/kernel_registry_manager.h

```diff
@@ -17,46 +18,52 @@ class KernelRegistry;
 class OpKernel;
 class SessionState;
 
-enum class KernelRegistryPriority {
-  HighPriority,
-  LowPriority
-};
-
 // Kernel registries' manager.
 // There're 2 kinds of kernel registries with priority from high to low as below,
 // 1. Custom execution provider type specific kernel registries.
 // 2. common execution provider type specific kernel registries.
 // The 1st and 2nd ones are shared across sessions.
+
+//This class is thread safe
 class KernelRegistryManager {
  public:
   KernelRegistryManager() = default;
 
-  void RegisterKernels(const ExecutionProviders& execution_providers,
-                       KernelRegistryPriority priority = KernelRegistryPriority::LowPriority);
+  void RegisterKernels(const ExecutionProviders& execution_providers);
 
```

Comment:
Does this create a more confusing setup? You now have things registered by RegisterKernelRegistry and this method in separate members. Based on the current implementation, values in kernel_registries_ will by higher priority than anything registered here. 

Why do the two sets of KernelRegistry instances need to be separate?

Reply:
They need to have different priorities. Currently, all the custom registries have higher priority than any provider. So, it would happen that:
1. user registered a kernel for GEMM op, but the kernel is only run on CPU
2. However, he prefer to use GPU, he registered CUDA provider before the CPU provider
3.  As the priority is defined as this, the cpu one will be used.

Reply:
Inside the implementation, we may not need two collections, just one is enough.

Reply:
Let me think about it.

Reply:
I didn't change the behavior. It still works exactly as before. 
The kernels either are from a provider, or from a custom registry. 
custom registries always have higher priority than the builtin ones.

Reply:
Would it be clearer if kernel_registries_ was called custom_kernel_registries_ and there was a comment before the member declaration explaining why that's separate from the values in provider_type_to_registry_ and how the two different collections are used? This is mainly to help developers understand the structure of the class. 

---
In reply to: [256193230](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r256193230) [](ancestors = 256193230)